File: 1301-1400/1305-All Elements in Two Binary Search Trees/1305-All Elements in Two Binary Search Trees.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<int> getAllElements(TreeNode* root1, TreeNode* root2) {
        stack<TreeNode*> St1, St2;
        pushLeft(St1, root1);
        pushLeft(St2, root2);
        vector<int> result;
        while (!St1.empty() || !St2.empty()) {
            stack<TreeNode*>& St = (St2.empty() || (!St1.empty() && St1.top()->val <= St2.top()->val)) ? St1 : St2;
            TreeNode* node = St.top();
            result.push_back(node->val);
            St.pop();
            pushLeft(St, node->right);
        }
        
        return result;
    }
    
private:
    void pushLeft(stack<TreeNode*>& St, TreeNode* node) {
        if (node) {
            St.push(node);
            pushLeft(St, node->left);
        }
    }
};

File: 1301-1400/1305-All Elements in Two Binary Search Trees/1305-All Elements in Two Binary Search Trees.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution:
    def getAllElements(self, root1: TreeNode, root2: TreeNode) -> List[int]:
        def pushLeft(St, node):
            if node:
                St.append(node)
                pushLeft(St, node.left)

        St1 = []
        St2 = []
        pushLeft(St1, root1)
        pushLeft(St2, root2)
        result = []
        while St1 or St2:
            St = St1 if not St2 or (St1 and St1[-1].val <= St2[-1].val) else St2
            result.append(St[-1].val)
            node = St.pop()
            pushLeft(St, node.right)
        return result
